clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3280,17 +3291,44 @@ void Sema::AddSYCLIntelNumSimdWorkItemsAttr(Decl *D,
       }
     }
 
-    // If the declaration has an [[intel::reqd_work_group_size]] attribute,
-    // check to see if the first argument can be evenly divided by the
-    // num_simd_work_items attribute.
+    // If the declaration has a [[intel::reqd_work_group_size()]]
+    // or [[cl::reqd_work_group_size()]] attribute, check to see
+    // if the last argument can be evenly divided by the
+    // [[intel::num_simd_work_items()]] attribute.
+    // If the declaration has a __attribute__((reqd_work_group_size))
+    // attribute, check to see if the first argument on OpenCL mode and
+    // last argument on SYCL mode can be evenly divided by the
+    // [[intel::num_simd_work_items()]] attribute.
     if (const auto *DeclAttr = D->getAttr<ReqdWorkGroupSizeAttr>()) {
-      Optional<llvm::APSInt> XDimVal = DeclAttr->getXDimVal(Context);
+      Expr *XDimExpr = DeclAttr->getXDim();
+      Expr *YDimExpr = DeclAttr->getYDim();
+      Expr *ZDimExpr = DeclAttr->getZDim();
+
+      if (!XDimExpr->isValueDependent() && !YDimExpr->isValueDependent() &&
+          !ZDimExpr->isValueDependent()) {
+        llvm::APSInt XDimVal, YDimVal, ZDimVal;
+        ExprResult XDim = VerifyIntegerConstantExpression(XDimExpr, &XDimVal);
+        ExprResult YDim = VerifyIntegerConstantExpression(YDimExpr, &YDimVal);
+        ExprResult ZDim = VerifyIntegerConstantExpression(ZDimExpr, &ZDimVal);
+
+        if (XDim.isInvalid())
```

Comment:
I'd combine these into a single `if` statement -- there's no need for `YDim` as that's otherwise unused.

Reply:
I sort of wonder whether we'd be doing a good thing by having an `AdditionalMember` like:
```
const Expr *getWorkGroupSize(const LangOptions &LO) const {
  return LO.OpenCL ? getXDim() : getZDim();
}
```
Perhaps not as part of this PR though -- this may be worth thinking about for when we correct the behavior of the one, two, and three arg forms of the attribute.

Reply:
> I sort of wonder whether we'd be doing a good thing by having an `AdditionalMember` like:
> 
> ```
> const Expr *getWorkGroupSize(const LangOptions &LO) const {
>   return LO.OpenCL ? getXDim() : getZDim();
> }
> ```
> 
> Perhaps not as part of this PR though -- this may be worth thinking about for when we correct the behavior of the one, two, and three arg forms of the attribute.

I agree.. I have created https://github.com/intel/llvm/issues/3761 so that we do not forget this. Thanks for pointing this.